### PR TITLE
Reuse the application's event loop group and thread pool when loading .env files

### PR DESCRIPTION
**Pull request by siemensikkema**
- Prevents a separate `EventLoopGroup` from being created unnecessarily for loading .env files.
- Deprecates overloads for `DotEnvFile.load` in favor of version that accept a `NonBlockingFileIO` so existing resources can be reused.
- Prevents multiple `EventLoopGroup`s from being created when calling the `load` methods that take an `Environment` when provided without a shared `EventLoopGroupProvider`.

### CODE REVIEW

**Comment by siemensikkema** [comment]
This is another attempt at fixing #2502. I think this comes very close to how .env files were loaded before Vapor 4.30.0. This avoids creating an extra `EventLoopGroup` and also reuses the application's threadpool (via Application.fileio).

I also ensured calling `DotEnvFile.load(for:on:fileio:logger:)` does not result in an EventLoopGroup for each file being loaded.

To avoid breaking the public API I kept the existing functions and made them call into the new overloads after creating a `NonBlockingFileIO`. I'm not so happy with the namespace pollution and code duplication but this is the best I could come up with for now.

File: Sources/Vapor/Utilities/DotEnv.swift [hunk 1]
```diff
@@ -42,26 +42,38 @@ public struct DotEnvFile {
...
+        defer {
+            do {
+                try threadPool.syncShutdownGracefully()
+            } catch {
+                logger.error("Shutting down threadPool failed: \(error)")
+            }
+        }
+        let fileio = NonBlockingFileIO(threadPool: threadPool)
+
+        self.load(for: environment, on: eventLoopGroupProvider, fileio: fileio, logger: logger)
```
Comment (MrLotU) [inline Sources/Vapor/Utilities/DotEnv.swift:56]: Can't put a comment on the function declaration, but we probably want to deprecate this one, correct?

File: Sources/Vapor/Utilities/DotEnv.swift [hunk 2]
```diff
@@ -86,19 +98,86 @@ public struct DotEnvFile {
...
+    ///     let elgp: EventLoopGroupProvider
+    ///     let logger: Logger
+    ///     try DotEnvFile.load(path: path, on: elgp, logger: logger)
+    ///     print(Environment.process.FOO) // BAR
+    ///
+    /// - parameters:
+    ///     - path: Absolute or relative path of the dotenv file.
+    ///     - eventLoopGroupProvider: Either provides an EventLoopGroup or tells the function to create a new one.
+    ///     - logger: Optionally provide an existing logger.
+    public static func load(
```
Comment (MrLotU) [inline Sources/Vapor/Utilities/DotEnv.swift:118]: Same here (deprecate?)

**Review by MrLotU — COMMENTED** [review]
LGTM overall. Just a question regarding deprecating the "faulty" methods in favor of the new ones. Since removing is indeed not an option.

**Comment by siemensikkema** [comment]
@MrLotU Good suggestion, I deprecated those functions.

**Review by MrLotU — APPROVED** [review]
LGTM! :D

**Review by JaapWijnen — APPROVED** [review]
LGTM, was thinking, can't we update the existing functions and add in the fileio with a default value. But that would stimulate people to use it in a way we want to prevent with this exact PR :P So deprecating the old and providing this new func looks like the way to go!

**Comment by siemensikkema** [comment]
@JaapWijnen I considered that but we can't use the `Application`s `NonBlockingFileIO` or `ThreadPool` as default arguments in these functions nor can we create a new `ThreadPool` in there because it has to be shut down properly.

**Comment by tanner0101** [comment]
These changes are now available in [4.34.0](https://github.com/vapor/vapor/releases/tag/4.34.0)

**Pull request merged**